android/src/main/java/com/tailscale/ipn/MainActivity.kt

```diff
@@ -285,6 +293,34 @@ class MainActivity : ComponentActivity() {
     lifecycleScope.launch { Notifier.loginFinished.collect { _ -> loginQRCode.set(null) } }
   }
 
+  private fun showPermissionDeniedDialog() {
+    AlertDialog.Builder(this)
+        .setTitle(R.string.vpn_permission_denied)
+        .setMessage(R.string.multiple_vpn_explainer)
+        .setPositiveButton(R.string.go_to_settings) { _, _ ->
+          // Intent to open the VPN settings
+          val intent = Intent(Settings.ACTION_VPN_SETTINGS)
+          startActivity(intent)
+        }
+        .setNegativeButton(R.string.cancel, null)
+        .show()
+  }
+
+  fun isAnotherVpnActive(context: Context): Boolean {
```

Comment:
Future: Is there a way to move this logic out of the main activity to prevent it from growing infinitely? On iOS, we have a dedicated, centralized class called ConflictDetector to perform this kind of system environment checks, instead of adding logic in different places.